src/microbe_stage/systems/FluidCurrentsSystem.cs

```diff
@@ -73,34 +87,84 @@ public FluidCurrentsSystem(float currentsTimePassed) : base(TemporarySystemHelpe
         noiseCurrentsY = null!;
     }
 
+    public void SetWorld(GameWorld world)
+    {
+        gameWorld = world;
+    }
+
     public Vector2 VelocityAt(Vector2 position)
     {
-        var scaledPosition = position * POSITION_SCALING;
+        if (!imagesInitialized)
+            return Vector2.Zero;
+
+        // This function's formula should be the same as the one in CurrentsParticles.gdshader
+        var scaledPosition = position * POSITION_SCALING * scale;
 
-        float disturbancesX = noiseDisturbancesX.GetNoise(scaledPosition.X, scaledPosition.Y,
-            currentsTimePassed * DISTURBANCE_TIMESCALE);
-        float disturbancesY = noiseDisturbancesY.GetNoise(scaledPosition.X, scaledPosition.Y,
-            currentsTimePassed * DISTURBANCE_TIMESCALE);
+        float disturbancesX = GetPixel(scaledPosition.X, scaledPosition.Y,
+            currentsTimePassed * DISTURBANCE_TIMESCALE * chaoticness, noiseDisturbancesXImage);
+        float disturbancesY = GetPixel(scaledPosition.X, scaledPosition.Y,
+            currentsTimePassed * DISTURBANCE_TIMESCALE * chaoticness, noiseDisturbancesYImage);
 
-        float currentsX = noiseCurrentsX.GetNoise(scaledPosition.X * CURRENTS_STRETCHING_MULTIPLIER,
-            scaledPosition.Y, currentsTimePassed * CURRENTS_TIMESCALE);
-        float currentsY = noiseCurrentsY.GetNoise(scaledPosition.X,
+        float currentsX = GetPixel(scaledPosition.X * CURRENTS_STRETCHING_MULTIPLIER,
+            scaledPosition.Y, currentsTimePassed * CURRENTS_TIMESCALE * chaoticness, noiseCurrentsXImage);
+        float currentsY = GetPixel(scaledPosition.X,
             scaledPosition.Y * CURRENTS_STRETCHING_MULTIPLIER,
-            currentsTimePassed * CURRENTS_TIMESCALE);
+            currentsTimePassed * CURRENTS_TIMESCALE * chaoticness, noiseCurrentsYImage);
 
         var disturbancesVelocity = new Vector2(disturbancesX, disturbancesY);
         var currentsVelocity = new Vector2(Math.Abs(currentsX) > MIN_CURRENT_INTENSITY ? currentsX : 0.0f,
             Math.Abs(currentsY) > MIN_CURRENT_INTENSITY ? currentsY : 0.0f);
 
-        return (disturbancesVelocity * DISTURBANCE_TO_CURRENTS_RATIO) +
-            (currentsVelocity * (1.0f - DISTURBANCE_TO_CURRENTS_RATIO));
+        return currentsVelocity.Lerp(disturbancesVelocity, DISTURBANCE_TO_CURRENTS_RATIO) * speed;
     }
 
     protected override void PreUpdate(float delta)
     {
         base.PreUpdate(delta);
 
+        if (!imagesInitialized)
+        {
+            var disturbancesX = noiseDisturbancesX.GetData();
```

Comment:
Can't this return null before the noise is generated? I think I had to take that into account in the thermal heat system (so the system needs to retry reading the data on each PreUpdate (and maybe skip running) until the noise image data is ready).